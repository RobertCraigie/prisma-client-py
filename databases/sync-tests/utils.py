from __future__ import annotations

import os
from pydantic import BaseModel

from .._types import DatabaseMapping
from .._compat import LiteralString


CURRENT_DATABASE = os.environ['PRISMA_DATABASE']


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
