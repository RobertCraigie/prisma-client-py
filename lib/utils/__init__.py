from ._utils import flatten as flatten
