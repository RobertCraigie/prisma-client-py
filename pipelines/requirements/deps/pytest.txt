pytest==7.4.4
