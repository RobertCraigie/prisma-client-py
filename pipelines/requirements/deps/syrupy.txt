syrupy==4.6.4
