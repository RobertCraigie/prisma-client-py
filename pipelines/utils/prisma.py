from __future__ import annotations

import nox


def generate(
    session: nox.Session,
    *,
    schema: str | None = 'tests/data/schema.prisma',
    clean: bool = True,
) -> None:
    if clean:
        session.run('python', '-m', 'prisma_cleanup')

    if schema:
        args = (f'--schema={schema}',)
    else:
        args = ()

    session.run('prisma', 'generate', *args)
