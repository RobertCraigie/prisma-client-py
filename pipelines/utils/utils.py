from pathlib import Path

import nox


CACHE_DIR = Path.cwd() / '.cache'


def get_pkg_location(session: nox.Session, pkg: str) -> str:
    location = session.run(
        'python',
        '-c',
        f'import {pkg}; print({pkg}.__file__)',
        silent=True,
    )
    assert isinstance(location, str)
    return str(Path(location).parent)


def setup_coverage(session: nox.Session, identifier: str) -> None:
    if identifier:
        identifier = f'.{identifier}'

    coverage_file = f'.coverage.{session.name}{identifier}'
    session.env['COVERAGE_FILE'] = str(CACHE_DIR / coverage_file)


def setup_env(session: nox.Session) -> None:
    setup_coverage(
        session,
        identifier=session.python if isinstance(session.python, str) else '',
    )
    session.env['PRISMA_PY_DEBUG'] = '1'
    session.env['PYTEST_ADDOPTS'] = ' '.join(
        [
            f'"{opt}"'
            for opt in [
                '-W error',
                # httpx deprecation warnings
                "-W ignore:'cgi' is deprecated and slated for removal in Python 3.13:DeprecationWarning",
                '-W ignore:path is deprecated:DeprecationWarning',
            ]
        ]
    )
