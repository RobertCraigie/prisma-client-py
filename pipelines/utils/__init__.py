from ._lark import generate_parsers as generate_parsers
from .utils import *
