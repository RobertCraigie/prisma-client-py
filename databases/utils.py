from __future__ import annotations

import os
from typing import Set
from pathlib import Path
from typing_extensions import Literal, get_args, override

from pydantic import BaseModel
from syrupy.location import PyTestLocation
from syrupy.extensions.amber import AmberSnapshotExtension

from ._types import DatabaseMapping
from ._compat import LiteralString

DatabaseFeature = Literal[
    'enum',
    'json',
    'date',
    'arrays',
    'array_push',
    'json_arrays',
    'raw_queries',
    'create_many_skip_duplicates',
    'transactions',
    'case_sensitivity',
    'full_text_search',
]


class DatabaseConfig(BaseModel):
    id: str
    name: str
    env_var: str
    bools_are_ints: bool
    autoincrement_id: str
    unsupported_features: Set[DatabaseFeature]
    default_date_func: str

    # TODO: run this under coverage
    def supports_feature(self, feature: DatabaseFeature) -> bool:  # pragma: no cover
        if feature not in get_args(DatabaseFeature):
            raise RuntimeError(f'Unknown feature: {feature}')

        return feature not in self.unsupported_features


# ------------------ Test helpers ------------------

from .constants import TESTS_DIR, SYNC_TESTS_DIR

SHARED_SNAPSHOTS_DIR = Path(__file__).parent.joinpath('__shared_snapshots__')
CURRENT_DATABASE = os.environ.get('PRISMA_DATABASE')


class AmberSharedExtension(AmberSnapshotExtension):
    """Syrupy extension that stores the snapshots in a parent __shared_snapshots__ dir"""

    @classmethod
    @override
    def dirname(cls, *, test_location: PyTestLocation) -> str:
        test_dir = Path(test_location.filepath).parent
        if test_dir.is_relative_to(SYNC_TESTS_DIR):
            rel_dir = test_dir.relative_to(SYNC_TESTS_DIR)
        else:
            rel_dir = test_dir.relative_to(TESTS_DIR)

        return str(SHARED_SNAPSHOTS_DIR.joinpath(rel_dir))


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
