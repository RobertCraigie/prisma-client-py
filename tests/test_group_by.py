import pytest
from syrupy.assertion import SnapshotAssertion

import prisma
from prisma import Prisma
from prisma.types import SortOrder

from .utils import async_fixture


# TODO: test all types
# TODO: test working with the results


@async_fixture(autouse=True, scope='session')
async def create_test_data(client: Prisma) -> None:
    create = client.profile.create
    await create(
        {
            'bio': 'from scotland',
            'country': 'Scotland',
            'city': 'Edinburgh',
            'views': 250,
            'user': {'create': {'name': 'Tegan'}},
        }
    )

    for _ in range(12):
        await create(
            {
                'bio': 'bio',
                'country': 'Denmark',
                'views': 500,
                'user': {'create': {'name': 'Robert'}},
            }
        )

    for _ in range(8):
        await create(
            {
                'bio': 'bio',
                'country': 'Denmark',
                'city': 'Copenhagen',
                'views': 1000,
                'user': {'create': {'name': 'Robert'}},
            }
        )

    types_create = client.types.create
    for i in range(10):
        await types_create(
            {
                'integer': i,
            },
        )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_group_by(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Basic test grouping by 1 field with no additional filters"""
    assert await client.user.group_by(['name']) == snapshot
    assert await client.profile.group_by(['country']) == snapshot


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_docs_example(
    snapshot: SnapshotAssertion, client: Prisma
) -> None:
    """Test the example given in the Prisma documentation:
    https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#groupby
    """
    results = await client.profile.group_by(
        by=['country', 'city'],
        count={
            '_all': True,
            'city': True,
        },
        sum={
            'views': True,
        },
        order={
            'country': 'desc',
        },
        having={
            'views': {
                '_avg': {
                    'gt': 200,
                },
            },
        },
    )
    assert results == snapshot


@pytest.mark.asyncio
@pytest.mark.persist_data
@pytest.mark.parametrize('order', ['asc', 'desc'])
async def test_order(
    snapshot: SnapshotAssertion, client: Prisma, order: SortOrder
) -> None:
    """Test ordering results by a grouped field"""
    assert (
        await client.profile.group_by(['country'], order={'country': order})
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_order_list(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Test ordering results by a list of grouped fields"""
    results = await client.profile.group_by(
        by=['country', 'city'],
        order=[
            {'country': 'asc'},
            {'city': 'desc'},
        ],
    )
    assert results == snapshot


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_order_multiple_fields(client: Prisma) -> None:
    """Test ordering results by multiple fields is not support"""
    with pytest.raises(prisma.errors.DataError):
        await client.profile.group_by(
            ['country', 'city'], order={'country': 'asc', 'city': 'desc'}
        )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_order_mismatched_arguments(client: Prisma) -> None:
    """The order argument only accepts fields that are being grouped"""
    with pytest.raises(prisma.errors.InputError) as exc:
        await client.profile.group_by(
            ['city'],
            order={  # pyright: reportGeneralTypeIssues=false
                'country': 'asc',
            },
        )

    assert exc.match(
        r'Every field used for orderBy must be included in the by-arguments of the query\. '
        r'Missing fields: country'
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
@pytest.mark.parametrize('order', ['asc', 'desc'])
async def test_take(
    snapshot: SnapshotAssertion, client: Prisma, order: SortOrder
) -> None:
    """Take argument limits number of records returned"""
    assert (
        await client.profile.group_by(
            ['country'], take=1, order={'country': order}
        )
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_take_missing_order_argument(client: Prisma) -> None:
    """The order argument must be provided to use take"""
    with pytest.raises(TypeError) as exc:
        await client.profile.group_by(['country'], take=1)

    assert exc.match(
        "Missing argument: 'order' which is required when 'take' is present"
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
@pytest.mark.parametrize('order', ['asc', 'desc'])
async def test_skip(
    snapshot: SnapshotAssertion, client: Prisma, order: SortOrder
) -> None:
    """Skipping grouped records"""
    assert (
        await client.profile.group_by(
            ['country'], skip=1, order={'country': order}
        )
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_skip_missing_order_argument(client: Prisma) -> None:
    """The order argument must be provided to use skip"""
    with pytest.raises(TypeError) as exc:
        await client.profile.group_by(['country'], skip=1)

    assert exc.match(
        "Missing argument: 'order' which is required when 'skip' is present"
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_where(client: Prisma) -> None:
    """Where argument correctly filters records"""
    results = await client.profile.group_by(
        ['country'], where={'country': 'Denmark'}
    )
    assert len(results) == 1
    assert results[0].get('country') == 'Denmark'

    results = await client.profile.group_by(
        ['country'], where={'bio': {'contains': 'scotland'}}
    )
    assert len(results) == 1
    assert results[0].get('country') == 'Scotland'


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_having_missing_field_in_by(client: Prisma) -> None:
    """Having filters must be an aggregation filter or be included in by"""
    with pytest.raises(prisma.errors.InputError) as exc:
        await client.profile.group_by(
            by=['country'],
            count=True,
            having={
                'views': {
                    'gt': 50,
                },
            },
        )

    assert exc.match(
        'Input error. Every field used in `having` filters must either be an aggregation filter '
        'or be included in the selection of the query. Missing fields: views'
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_having_aggregation(
    snapshot: SnapshotAssertion, client: Prisma
) -> None:
    """Having aggregation filters records correctly"""
    assert (
        await client.profile.group_by(
            by=['country'],
            count=True,
            having={
                'views': {
                    '_avg': {
                        'gt': 600,
                    }
                }
            },
        )
        == snapshot
    )
    assert (
        await client.profile.group_by(
            by=['country'],
            count=True,
            having={
                'views': {
                    '_avg': {
                        'lt': 600,
                    }
                }
            },
        )
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_having_aggregation_nested(
    snapshot: SnapshotAssertion, client: Prisma
) -> None:
    """Having aggregation filters nested within statements correctly filters records"""
    results = await client.profile.group_by(
        by=['country'],
        count=True,
        having={
            'OR': [
                {
                    'views': {
                        '_avg': {
                            'equals': 1000,
                        },
                    },
                },
                {
                    'views': {
                        '_sum': {
                            'equals': 250,
                        },
                    },
                },
            ],
        },
    )
    assert results == snapshot

    results = await client.profile.group_by(
        by=['country'],
        count=True,
        having={
            'OR': [
                {
                    'views': {
                        '_avg': {
                            'equals': 700,
                        },
                    },
                },
                {
                    'views': {
                        '_sum': {
                            'equals': 250,
                        },
                    },
                },
            ],
        },
    )
    assert results == snapshot

    results = await client.profile.group_by(
        by=['country'],
        count=True,
        having={
            'OR': [
                {
                    'views': {
                        '_avg': {
                            'equals': 700,
                        },
                    },
                },
                {
                    'views': {
                        '_sum': {
                            'equals': 250,
                        },
                    },
                    'NOT': [
                        {
                            'views': {
                                '_min': {
                                    'equals': 250,
                                },
                            },
                        },
                    ],
                },
            ],
        },
    )
    assert results == snapshot


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_count(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Counting records"""
    assert await client.profile.group_by(['country'], count=True) == snapshot
    assert (
        await client.profile.group_by(['country'], count={'_all': True})
        == snapshot
    )
    assert (
        await client.profile.group_by(['country'], count={'city': True})
        == snapshot
    )
    assert (
        await client.profile.group_by(
            ['country'], count={'city': True, 'country': True}
        )
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_avg(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Getting the average of records"""
    assert (
        await client.profile.group_by(['country'], avg={'views': True})
        == snapshot
    )
    assert (
        await client.types.group_by(
            ['string'], avg={'integer': True, 'bigint': True}
        )
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_sum(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Getting the sum of records"""
    assert (
        await client.profile.group_by(['country'], sum={'views': True})
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_min(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Getting the minimum value of records"""
    assert (
        await client.profile.group_by(['country'], min={'views': True})
        == snapshot
    )


@pytest.mark.asyncio
@pytest.mark.persist_data
async def test_max(snapshot: SnapshotAssertion, client: Prisma) -> None:
    """Getting the maximum value of records"""
    assert (
        await client.profile.group_by(['country'], max={'views': True})
        == snapshot
    )
