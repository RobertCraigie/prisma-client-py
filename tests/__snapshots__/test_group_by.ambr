# name: test_avg
  list([
    dict({
      '_avg': dict({
        'views': 700.0,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_avg': dict({
        'views': 250.0,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_avg.1
  list([
    dict({
      '_avg': dict({
        'bigint': 0.0,
        'integer': 4.5,
      }),
      'string': '',
    }),
  ])
# ---
# name: test_count
  list([
    dict({
      '_count': dict({
        '_all': 20,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_count': dict({
        '_all': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_count.1
  list([
    dict({
      '_count': dict({
        '_all': 20,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_count': dict({
        '_all': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_count.2
  list([
    dict({
      '_count': dict({
        'city': 8,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_count': dict({
        'city': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_count.3
  list([
    dict({
      '_count': dict({
        'city': 8,
        'country': 20,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_count': dict({
        'city': 1,
        'country': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_docs_example
  list([
    dict({
      '_count': dict({
        '_all': 1,
        'city': 1,
      }),
      '_sum': dict({
        'views': 250,
      }),
      'city': 'Edinburgh',
      'country': 'Scotland',
    }),
    dict({
      '_count': dict({
        '_all': 12,
        'city': 0,
      }),
      '_sum': dict({
        'views': 6000,
      }),
      'city': None,
      'country': 'Denmark',
    }),
    dict({
      '_count': dict({
        '_all': 8,
        'city': 8,
      }),
      '_sum': dict({
        'views': 8000,
      }),
      'city': 'Copenhagen',
      'country': 'Denmark',
    }),
  ])
# ---
# name: test_group_by
  list([
    dict({
      'name': 'Robert',
    }),
    dict({
      'name': 'Tegan',
    }),
  ])
# ---
# name: test_group_by.1
  list([
    dict({
      'country': 'Denmark',
    }),
    dict({
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_having_aggregation
  list([
    dict({
      '_count': dict({
        '_all': 20,
      }),
      'country': 'Denmark',
    }),
  ])
# ---
# name: test_having_aggregation.1
  list([
    dict({
      '_count': dict({
        '_all': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_having_aggregation_nested
  list([
    dict({
      '_count': dict({
        '_all': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_having_aggregation_nested.1
  list([
    dict({
      '_count': dict({
        '_all': 20,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_count': dict({
        '_all': 1,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_having_aggregation_nested.2
  list([
    dict({
      '_count': dict({
        '_all': 20,
      }),
      'country': 'Denmark',
    }),
  ])
# ---
# name: test_max
  list([
    dict({
      '_max': dict({
        'views': 1000,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_max': dict({
        'views': 250,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_min
  list([
    dict({
      '_min': dict({
        'views': 500,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_min': dict({
        'views': 250,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_order[asc]
  list([
    dict({
      'country': 'Denmark',
    }),
    dict({
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_order[desc]
  list([
    dict({
      'country': 'Scotland',
    }),
    dict({
      'country': 'Denmark',
    }),
  ])
# ---
# name: test_order_list
  list([
    dict({
      'city': 'Copenhagen',
      'country': 'Denmark',
    }),
    dict({
      'city': None,
      'country': 'Denmark',
    }),
    dict({
      'city': 'Edinburgh',
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_skip[asc]
  list([
    dict({
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_skip[desc]
  list([
    dict({
      'country': 'Denmark',
    }),
  ])
# ---
# name: test_sum
  list([
    dict({
      '_sum': dict({
        'views': 14000,
      }),
      'country': 'Denmark',
    }),
    dict({
      '_sum': dict({
        'views': 250,
      }),
      'country': 'Scotland',
    }),
  ])
# ---
# name: test_take[asc]
  list([
    dict({
      'country': 'Denmark',
    }),
  ])
# ---
# name: test_take[desc]
  list([
    dict({
      'country': 'Scotland',
    }),
  ])
# ---
